### PR TITLE
show all lines of an error. #150

### PR DESCRIPTION
**Pull request by coreyoconnor**
This has an issue where the error line order might output incorrectly. Not skilled in emacs/elisp so any suggestions would be great :)

### CODE REVIEW

**Comment by coreyoconnor** [comment]
Here is a contrived example of a multi-line error message displayed using this change:

image

Without this change only the "scalac: missing argument list for method run in class IndexedReaderWriterStateT" text is displayed.

**Comment by sebastiencs** [comment]
Thanks @coreyoconnor, could you add a `defcustom` to enable this behavior, as suggested by @yyoncho.

The issue with this change is that it could display lines of an error in disorder.
`lsp-ui-sideline--find-line` will return the first line (position) which has enough space to display that line.

For instance, if an error is:
```
..........First line of an error..........
Second line
```

It could be display like this:
```
Second line
..........First line of an error..........
```

To fix this you need to check that `lsp-ui-sideline--find-line` returns always a number (position) superior to the one in the previous iteration

**Comment by nasyxx** [comment]
So glad that this issue can be solved.  Thank you very much. 

For a long time, when I used Haskell, I closed the sideline's flycheck because it displays too few and even only one "error".

**Comment by coreyoconnor** [comment]
@nasyxx have you tried these changes with Haskell? I've only tried Scala. As @sebastiencs noted there is an issue with ordering but otherwise should work.

Should be easy to fix... but my elisp time is up for a bit :)

**Comment by nasyxx** [comment]
I have tried it; it did the fix.

screen shot 2019-01-30 at 8 13 37 am

The old one only has the message which inside `[]`.

**Comment by coreyoconnor** [comment]
I added a defcustom. I won't be correcting the out-of-order aspect in the foreseeable future. (non critical)

The option `lsp-ui-sideline-diagnostic-all-lines` defaults to true. I don't know of a usecase where only showing the first line is useful. So that's reasonable to me.

That said, I think a  "maximum number of lines" seems a better option. What do you think? Default to something large. Would cover the single line case, the usual, and mitigate those servers that output too much. (Not sure which servers do that but I suspect C++ would haha)

Also, I'd prefer if this was per-mode but I'm unsure how to do that. defvar-local?

File: lsp-ui-sideline.el [hunk 1]
```diff
@@ -305,23 +312,33 @@ CURRENT is non-nil when the point is on the symbol."
...
+                   (message (progn (add-face-text-property 0 len 'lsp-ui-sideline-global nil message)
+                                   (add-face-text-property 0 len face nil message)
+                                   message))
+                   (string (concat (propertize " " 'display `(space :align-to (- right-fringe ,(lsp-ui-sideline--align len margin))))
+                                   message))
+                   (pos-ov (lsp-ui-sideline--find-line len nil))
+                   (ov (and pos-ov (make-overlay pos-ov pos-ov))))
+              (when pos-ov
+                (overlay-put ov 'after-string string)
+                (push ov lsp-ui-sideline--ovs)))
```
Comment (yyoncho) [inline lsp-ui-sideline.el]: please place these on one line

Comment (yyoncho) [reply lsp-ui-sideline.el]: (I mean the brackets)

**Comment by coreyoconnor** [comment]
Will do!

File: lsp-ui-sideline.el (hunk 1 above)
Comment (coreyoconnor) [reply lsp-ui-sideline.el]: Is there a reference for the style? I think I corrected this.

**Comment by coreyoconnor** [comment]
@yyoncho Can you review again? Instead of a boolean the change now introduces a `lsp-ui-sideline-diagnostic-max-lines` integer variable.

File: lsp-ui-sideline.el [hunk 2]
```diff
@@ -306,24 +313,35 @@ CURRENT is non-nil when the point is on the symbol."
...
+                   (string (concat (propertize " " 'display `(space :align-to (- right-fringe ,(lsp-ui-sideline--align len margin))))
+                                   message))
+                   (pos-ov (lsp-ui-sideline--find-line len nil))
+                   (ov (and pos-ov (make-overlay pos-ov pos-ov)))
+                   )
+              (when pos-ov
+                (overlay-put ov 'after-string string)
+                (overlay-put ov 'kind 'diagnotics)
+                (push ov lsp-ui-sideline--ovs)
+                )
```
Comment (yyoncho) [inline lsp-ui-sideline.el]: Can you put all the closing brackets on one line?

Comment (coreyoconnor) [reply lsp-ui-sideline.el]: Done. I think? I'm ignorant of emacs style and style tooling.

**Comment by coreyoconnor** [comment]
@sebastiencs 

> To fix this you need to check that `lsp-ui-sideline--find-line` returns always a number (position) superior to the one in the previous iteration

Completed a first pass at this. `lsp-ui-sideline--find-line` takes an optional parameter `offset` that is incremented each iteration. The function now returns cons `pos` and `index` which is used to update offset. Please review. :)

**Comment by coreyoconnor** [comment]
This is complete. All recent changes are rebases. Please merge :)

**Pull request merged**